grafana/grafana
# CloudWatch: Remove simplejson in favor of 'encoding/json'

**Which issue(s) this PR fixes**:
The `simplejson` library is [no longer recommended by the backend team](https://github.com/grafana/grafana/blob/main/contribute/style-guides/backend.md?plain=1#L189). Therefore the Cloud datasources team has decided to factor the library out of our plugins in favor of the `encoding/json` package that's part of the Go standard library.

Fixes #45417

**Special notes for your reviewer**:
`simplejson` provides several helper methods to query data and to set defaults if the data returns "empty", e.g. `parameters.Get("limit").MustInt64(50)` gets the `limit` property and if it's `null` sets the value to `50`. Using the `encoding/json` package for this results in a few lines of code instead of one, e.g. 

```go
region := "some-default"
if model.Region != "" {
  region = model.Region
}
```

The second aspect of this is where `simplejson` will return an `Error` if the property isn't found, e.g. `_, err := queryJson.Get("statistic").String()`. Using the `encoding/json` package for this means we need to modify our default struct definition to be a pointer to a type, e.g. `*string` instead of `string`. Here's a [playground link](https://play.golang.com/p/r1CmOG5tCF2) showing the difference in `Unmarshal`.

**Verification**
1. Run `go test` in the `cloudwatch` package, all test should pass.
2. Visit any of the premade dashboards for the Cloudwatch plugin, they should work without errors.
